### PR TITLE
rough draft of orders transaction fact table incremental model & DDL …

### PR DESCRIPTION
**Pull request by Jon-Simpson**
…for source tables

### CODE REVIEW

File: stonecrop_data_warehouse/ddl/sources/capstan.sql [hunk 1]
```diff
@@ -0,0 +1,3817 @@
+BEGIN;
+
+DROP SCHEMA IF EXISTS sources CASCADE;
+CREATE SCHEMA sources;
+
+CREATE TABLE sources.addresses (
+    xid bigint default null,
+    db_commit_time timestamp default null,
```
Comment (ezixi) [inline stonecrop_data_warehouse/ddl/sources/capstan.sql]: I've been using `timestamptz` as a default datatype for timestamps. We have multiple data sources and they have different default timezones (Capstan UI is in EST - i think - and the Capstan DB is in UTC; WMS is in central time and I have no clue what Atlas defaults to) so either we convert everything to UTC on ingest, or will carry the time zone info with the data. I've done the latter, but I could be persuaded to to do the former.

File: stonecrop_data_warehouse/ddl/sources/capstan.sql [hunk 2]
```diff
@@ -0,0 +1,3817 @@
+BEGIN;
```
Comment (ezixi) [inline stonecrop_data_warehouse/ddl/sources/capstan.sql]: Are these sources using the dbt concept of 'sources'? If so, I know it's a pain in the arse, but we might better having one file per table to help version control and maintenance in the future...

File: stonecrop_data_warehouse/ddl/sources/capstan.sql (hunk 1 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/ddl/sources/capstan.sql]: I like the rule of thumb to default to UTC in apps, but since the warehouse ought to be  adding meaning to data, I'd favor actually putting the proper timestamps on everything during staging. I'll update those

File: stonecrop_data_warehouse/ddl/sources/capstan.sql (hunk 2 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/ddl/sources/capstan.sql]: honestly, I just made that sources file because I saw a page in the docs about it. I can add those refs if you'd like, but I don't actually see a purpose in actually using them yet. I'm planning on those source tables being queried just once each. Although we may very well want to use them in the future. I like the idea of splitting them up though. same with my monster DDL file to create the source tables.. not quite there yet.

This whole PR is a very rough draft of things, so you're right about it not being ready for a "proper review" lol

File: stonecrop_data_warehouse/ddl/sources/capstan.sql (hunk 1 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/ddl/sources/capstan.sql]: Actually, I'm going to hold off on this change for now, since these are DDL for source tables, which probably shouldn't be imposing any meaning on the raw data. That's something that should definitely be done at the next staging step, though

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql [hunk 3]
```diff
@@ -85,31 +88,31 @@ select
...
-LEFT JOIN sources.ctd_boms ON ctd_boms.id = orders.site_augment_instance_id
-LEFT JOIN sources.ctd_site_designs ON ctd_site_designs.id = orders.site_augment_instance_id
\ No newline at end of file
+LEFT JOIN {{ ref('stg_capstan__phases') }} phases ON phases.id = orders.phase_id
+LEFT JOIN {{ ref('stg_capstan__distribution_centers') }} distribution_centers ON orders.distribution_center_id = distribution_centers.id
+LEFT JOIN {{ ref('stg_capstan__augments') }} augments ON augments.id = phases.augment_id
+-- join for augment_categories
+LEFT JOIN {{ ref('stg_capstan__work_orders') }} work_orders ON work_orders.order_id = orders.id
+-- join for ctd_site_augment_instances
+LEFT JOIN {{ ref('stg_capstan__issues') }} issues ON issues.origin_id = orders.id
```
Comment (bassehes) [inline stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: issues.origin_id is not unique - they have different origin_type's. For this would assume you want to specify origin_type` Order`.

Issues can also be opened and closed. I assume for this model you want to just show issues where the `closed_on` is null.

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql [hunk 4]
```diff
@@ -85,31 +88,31 @@ select
...
+-- join for issue_codes
+LEFT JOIN {{ ref('stg_capstan__milestones') }} milestones ON milestones.event_id = orders.order_id
```
Comment (bassehes) [inline stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: The `milestones` table is similar to `issues` where there is a type. Here would use `milestones.event_type` `Order`.

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql (hunk 3 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: Yeah, the idea here is that there are indeed many rows per order (similar to the milestones table), as an append only fact table will be faster and easier to construct. Exactly how we want this table to look is not vetted yet, so I'm trying to be inclusive here. But I'm think we can throw views on top of it if we wanted the "crosstab"ed look with just order milestones

Good point though

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql [hunk 5]
```diff
@@ -1,18 +1,21 @@
 
 select
 -- TODO: add fields relevant to replication events 
+-- event fields
+    transaction_type,
+
 -- orders
-    orders.id,
+--    orders.id,
```
Comment (ezixi) [inline stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: these look commented out, so not sure if this applies, but... I don't like having fields that are just 'id'. Yes, you can figure out what they represent from the table name / context, but i'd prefer that we be explicit and `select orders.id as order_id from orders;`

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql [hunk 6]
```diff
@@ -85,31 +88,31 @@ select
...
-LEFT JOIN sources.issues ON issues.origin_id = orders.id
-LEFT JOIN sources.issue_codes ON issue_codes.
-LEFT JOIN sources.milestones ON milestones.event_id = orders.order_id
-LEFT JOIN sources.locations ON locations.id = orders.location_id
-LEFT JOIN sources.markets ON markets.id = locations.market_id
-LEFT JOIN sources.regions ON region.id = markets.region_id
-LEFT JOIN sources.ctd_boms ON ctd_boms.id = orders.site_augment_instance_id
-LEFT JOIN sources.ctd_site_designs ON ctd_site_designs.id = orders.site_augment_instance_id
\ No newline at end of file
+LEFT JOIN {{ ref('stg_capstan__phases') }} phases ON phases.id = orders.phase_id
```
Comment (ezixi) [inline stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: a good by-product of explicitly naming the primary / foreign keys is that you can then join models like
```suggestion
LEFT JOIN {{ ref('stg_capstan__phases') }} phases using (phase_id)
```

which, I'm assured, is more performant.

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql (hunk 3 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: constraint added for 'Orders'

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql (hunk 4 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: Done.

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql (hunk 5 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: I agree. I commented out all fields that I wasn't sure if I wanted yet, but if/when they are uncommented, I'll follow that convention

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql [hunk 7]
```diff
@@ -85,31 +88,34 @@ select
...
+LEFT JOIN {{ ref('stg_capstan__issues') }} issues ON issues.origin_id = orders.id
+-- join for issue_codes
+LEFT JOIN {{ ref('stg_capstan__milestones') }} milestones ON milestones.event_id = orders.order_id
+LEFT JOIN {{ ref('stg_capstan__locations') }} locations ON locations.id = orders.location_id
+-- join for markets
+-- join for regions
+-- join for ctd_boms
+LEFT JOIN {{ ref('stg_capstan__ctd_site_designs') }} ctd_site_designs ON ctd_site_designs.id = orders.site_augment_instance_id
+WHERE
+    milestones.event_type = 'Order'
```
Comment (bassehes) [inline stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: If you don't do something like `(milestones.event_type = 'Order' or milestones.event_type is null)` won't it limit the results if there is no milestone for a given order (trying to think if that is even a possibility).

Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: I _think_ that's what we want here. Would need to test it out, and we aren't there yet, but I I wouldn't want to include a row here that has a null milestone, as that is not an event. Since this will be a fact table, it should just contain events ( i.e. significant changes) to an order

File: stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql (hunk 6 above)
Comment (Jon-Simpson) [reply stonecrop_data_warehouse/models/marts/business/fct_capstan__orders.sql]: Okay, I updated the JOINS which I were immediately applicable. Some don't join on columns of the same name (milestones uses event_id as order ID, etc)

File: stonecrop_data_warehouse/models/staging/capstan/capstan.yml [hunk 8]
```diff
@@ -181,18 +181,18 @@ models:
             - name: db_event
                 description: The type of database event this record represents (insert/update/delete)
             - name: issue_id
-                description:
+                description: Unique issue identifier. Primary key in relational model.
             - name: origin_id 
-                description:
+                description: "?"
```
Comment (mfinaly) [inline stonecrop_data_warehouse/models/staging/capstan/capstan.yml]: Is there a way to circle back on these, or tag them so others who might know can chime in?

Comment (ezixi) [reply stonecrop_data_warehouse/models/staging/capstan/capstan.yml]: That's a good point... I'm not sure..

Comment (ezixi) [reply stonecrop_data_warehouse/models/staging/capstan/capstan.yml]: I added a 'missing docs' label, but not sure if that is just additional noise...
